src/extension/src/coreTemplateStudio.ts

```diff
@@ -94,144 +84,143 @@ export class CoreTemplateStudio {
     }
   }
 
-  private constructor(process: ChildProcess, port: number, url: string) {
-    this._process = process;
-    this._port = port;
-    this._url = url;
+  private constructor(processCli: ChildProcess) {
+    this._processCli = processCli;
+    this.cliEvents = new CliEventEmitter();
+    this.readStream(this._processCli);
   }
 
-  public getPort(): number {
-    return this._port;
+  // This function is a listener, in the constructor, it gets attached
+  // then it will always get triggered when a command is write to the cli and there is responses from cli, until the process gets killed
+  public async readStream(process: ChildProcess) {
+    let data = "";
+    process.stdout.on("data", (chunk) => {
+      data += chunk;
+      var responses = data.toString().split("\n");
+      for (let i = 0; i < responses.length - 1; i++) {
+        var result = JSON.parse(responses[i]);
+        this.cliEvents.emit(result['messageType'], result['content']);
+      }
+      data = responses[responses.length - 1];
+    });
+
+    process.stderr.on("data", (data) => {
+      this.cliEvents.emit("error", data.toString());
+    });
   }
 
-  public async getProjectTypes(): Promise<any> {
-    // TODO: use this in client instead of fetching directly from API server
-    const url = new URL(CONSTANTS.API.ENDPOINTS.PROJECT_TYPE, this._url);
-    return await fetch(url.href, { method: CONSTANTS.API.METHODS.GET })
-      .then((response: Response) => {
-        return response.json();
-      })
-      .catch((error: Error) => {
-        throw Error(error.toString());
+  private async awaitCliEvent(eventName: string): Promise<any> {
+    return new Promise((resolve, reject) => {
+      this.cliEvents.once(eventName, (data) => {
+        this.cliEvents.removeAllListeners();
+        resolve(data);
+      });
+      this.cliEvents.once("error", (data) => {
+        this.cliEvents.removeAllListeners();
+        reject(data);
       });
+    });
   }
 
-  public async getFrameworks(projectType: string): Promise<any> {
-    const url = new URL(CONSTANTS.API.ENDPOINTS.FRAMEWORK, this._url);
-    url.searchParams.append(
-      CONSTANTS.API.QUERY_PARAMS.PROJECT_TYPE,
-      projectType
-    );
+  public async sync(payload: ICommandPayload): Promise<any> {
+    let syncCommand = `${CONSTANTS.CLI.SYNC_COMMAND_PREFIX} -p ${payload.payload.path}\n`;
+    this._processCli.stdin.write(syncCommand);
+    this.cliEvents.on(CONSTANTS.CLI.SYNC_PROGRESS_STATE, (data) => {
+      payload.liveMessageHandler(data['status'], data['progress']);
+    });
+    return await this.awaitCliEvent(CONSTANTS.CLI.SYNC_COMPLETE_STATE);
```

Comment:
Because you are returning a promise `await` is not needed here

Reply:
The reason is I want promise get resolved before it gets returned. Thus I use await here. Await puts an pause in the code.

Reply:
"await suspends the execution until an asynchronous function return promise is fulfilled and unwraps the value from the Promise returned. " from Typescipt docs

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-1-7.html

Reply:
You're still returning a promise, it has the same exact effect if you remove the await, I would incite you to test it!